File: docs/caching/method-dispatch-bench.js
#!/usr/bin/env node
"use strict";
/// Micro-benchmark: generic function dispatch (TypeScript version)
/// Using a dispatch table with type-based function selection
// Type checking function
function getType(x) {
    if (typeof x === "number")
        return "number";
    if (typeof x === "string")
        return "string";
    if (Array.isArray(x)) {
        if (typeof x[0] === "number")
            return "list";
        return "vector";
    }
    if (typeof x === "symbol")
        return "symbol";
    return "other";
}
const dispatchTable = {
    number: (x) => ["number", x],
    string: (x) => ["string", x.length],
    list: (x) => ["list", x.length],
    vector: (x) => ["vector", x.length],
    symbol: (x) => ["symbol", x],
    other: (x) => ["other", x],
};
// Generic dispatcher using dispatch table
function dispatchGeneric(x) {
    const type = getType(x);
    const fn = dispatchTable[type] || dispatchTable["other"];
    return fn(x);
}
// Test data: 2,000,000 calls with 5-type repeating cycle
const testData = [];
const cycle = [
    1,
    "test string",
    [1, 2, 3, 4, 5],
    [1, 2, 3, 4, 5],
    Symbol("symbol"),
];
for (let i = 0; i < 2000000; i++) {
    testData.push(cycle[i % cycle.length]);
}
// Benchmark function
function benchmark(iterations) {
    console.log("Running dispatch benchmark (3 iterations):");
    for (let run = 0; run < iterations; run++) {
        const startNs = process.hrtime.bigint();
        for (let i = 0; i < testData.length; i++) {
            const result = dispatchGeneric(testData[i]);
        }
        const elapsedNs = process.hrtime.bigint() - startNs;
        const elapsedMs = Number(elapsedNs) / 1000000;
        const elapsedSec = elapsedMs / 1000;
        console.log(`  Run ${run + 1}: ${elapsedSec.toFixed(1)} seconds`);
    }
}
// Main
console.log("================================");
console.log("TypeScript Generic Function Dispatch Micro-Benchmark");
console.log("================================");
console.log(`Implementation: TypeScript (Node.js ${process.version})`);
console.log("Test data: 2,000,000 calls over repeating 5-type cycle");
console.log("  Type cycle: number -> string -> list -> vector -> symbol\n");
console.log("Warming up JIT compiler (10,000 calls)...");
for (let i = 0; i < 10000; i++) {
    dispatchGeneric(testData[i % testData.length]);
}
console.log("Warmup complete.\n");
benchmark(3);
console.log("\n================================");
console.log("Benchmark Complete");
console.log("================================");

File: docs/caching/method_dispatch_bench.py
#!/usr/bin/env python3
"""
Micro-benchmark: generic function dispatch (Python version)
Basic dispatch performance test
"""

import time
from typing import Any, List, Callable

# Simple type-based dispatcher
dispatch_table = {}

def defmethod(type_name: str) -> Callable:
    """Decorator to register a dispatcher method"""
    def decorator(fn: Callable) -> Callable:
        dispatch_table[type_name] = fn
        return fn
    return decorator

@defmethod('long')
def dispatch_long(x):
    return ('long', x)

@defmethod('string')
def dispatch_string(x):
    return ('string', len(x))

@defmethod('list')
def dispatch_list(x):
    return ('list', len(x))

@defmethod('dict')
def dispatch_dict(x):
    return ('dict', len(x))

@defmethod('symbol')
def dispatch_symbol(x):
    return ('symbol', x)

@defmethod('other')
def dispatch_other(x):
    return ('other', x)

def dispatch_clj_impl(x):
    """Generic dispatcher using type-based dispatch"""
    if isinstance(x, (int, float)):
        key = 'long'
    elif isinstance(x, str):
        key = 'string'
    elif isinstance(x, list):
        key = 'list'
    elif isinstance(x, dict):
        key = 'dict'
    elif isinstance(x, tuple) and len(x) == 1 and isinstance(x[0], str):
        key = 'symbol'
    else:
        key = 'other'

    return dispatch_table[key](x)

# Test data: 200,000 calls with 5-type repeating cycle
test_data = []
cycle = [1, "test string", [1, 2, 3, 4, 5], {'a': 1}, ('symbol',)]
for i in range(200000):
    test_data.append(cycle[i % 5])

def benchmark_dispatch(iterations: int) -> List[int]:
    """Benchmark generic dispatcher"""
    times = []
    for run in range(iterations):
        result = 0
        start_ns = time.perf_counter()
        for item in test_data:
            r = dispatch_clj_impl(item)
            if r:
                result += 1
        elapsed_ms = (time.perf_counter() - start_ns) * 1000.0
        print(f"  Run {run + 1}: {elapsed_ms / 1000.0:.1f} seconds")
        times.append(result)
    return times

def run_all_benchmarks():
    """Run all benchmarks"""
    print("\n================================")
    print("Python Generic Function Dispatch Micro-Benchmark")
    print("================================")
    import sys
    print(f"Implementation: Python {sys.version.split()[0]}")
    print("Test data: 200,000 calls over repeating 5-type cycle")
    print("  Type cycle: int -> string -> list -> dict -> symbol\n")

    print("Warming up JIT compiler (10,000 calls)...")
    for i in range(10000):
        dispatch_clj_impl(test_data[i % len(test_data)])
    print("Warmup complete.\n")

    print("Running dispatch benchmark (3 iterations):")
    benchmark_dispatch(3)

    print("\n================================")
    print("Benchmark Complete")
    print("================================")

if __name__ == "__main__":
    run_all_benchmarks()
